File: source/benchmark.py
import time


def get_time_us():
    return time.perf_counter() * 1e6


def heavy_task(N):
    sum_v = 0.0
    for i in range(N):
        sum_v += i * 0.1234
    return sum_v


def benchmark():
    N = 50 * 1000 * 1000  # 50 million ops

    # Warm-up
    for _ in range(3):
        heavy_task(N // 10)

    total_us = 0.0
    test_runs = 10

    for i in range(test_runs):
        start = get_time_us()
        _ = heavy_task(N)
        end = get_time_us()

        elapsed = end - start
        total_us += elapsed

        print(f"Run {i+1:2d}: {elapsed/1000.0:.3f} ms ({elapsed:.0f} us)")

    avg = total_us / test_runs
    print(f"\nAverage: {avg/1000.0:.3f} ms ({avg:.0f} us)")


if __name__ == "__main__":
    benchmark()

File: source/benchmark.c
#include <stdio.h>
#include <stdint.h>

#if defined(_WIN32) || defined(_WIN64)
#include <windows.h>

double get_time_us()
{
    LARGE_INTEGER freq, counter;
    QueryPerformanceFrequency(&freq);
    QueryPerformanceCounter(&counter);
    return (double)counter.QuadPart * 1e6 / (double)freq.QuadPart;
}

#else
#include <time.h>

double get_time_us()
{
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (double)ts.tv_sec * 1e6 + (double)ts.tv_nsec / 1e3;
}

#endif

// Example workload: simple math loop
static inline double heavy_task(int N)
{
    double sum = 0.0;
    for (int i = 0; i < N; i++)
    {
        sum += i * 0.1234;
    }
    return sum;
}

int main()
{
    const int N = 50 * 1000 * 1000; // 50 million ops
    double start = get_time_us();

    double result = heavy_task(N);

    double end = get_time_us();
    double elapsed_us = end - start;

    printf("Task result: %.2f\n", result);
    printf("Elapsed: %.3f ms (%.0f us)\n", elapsed_us / 1000.0, elapsed_us);

    return 0;
}
